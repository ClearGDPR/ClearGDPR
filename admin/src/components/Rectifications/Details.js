import React from 'react';
import PropTypes from 'prop-types';
import lodash from 'lodash';

import { rectificationDetailsType } from 'types';

import Loader from 'components/core/cards/dashboard/Loader';
import { PrimaryButton } from 'components/core/Common/Buttons/Buttons';

const Details = ({ rectification, onApprove, isLoading }) => {
  function renderObject(obj, level = 0) {
    if (!obj) return;

    return Object.keys(obj).map((k, index) => (
      <React.Fragment key={index}>
        {level ? <br /> : ''}
        {[...Array(level)].map((value, arrayIndex) => (
          <React.Fragment key={arrayIndex}>&nbsp;&nbsp;</React.Fragment>
        ))}
        {k}:&nbsp;
        {lodash.isObject(obj[k]) ? renderObject(obj[k], level + 1) : obj[k]}
        {!level ? <br /> : ''}
      </React.Fragment>
    ));
  }

  function renderDetails() {
    return (
      <React.Fragment>
        <p>
          <strong>Created at:</strong> {rectification.created_at}
        </p>
        <p>
          <strong>Current data:</strong>
        </p>
        <p>{renderObject(rectification.currentData)}</p>
        <p>
          <strong>Updates:</strong>
        </p>
        <p>{renderObject(rectification.updates)}</p>
        <p>Status: {rectification.status}</p>
        {rectification.status !== 'APPROVED' && (
          <PrimaryButton
            type="button"
            onClick={e => {
              e.preventDefault();
              onApprove();
            }}
            text="Approve"
          />
        )}
      </React.Fragment>
    );
  }
  return isLoading ? <Loader /> : renderDetails();
};

Details.propTypes = {
  rectification: rectificationDetailsType.isRequired,
  onApprove: PropTypes.func.isRequired,
  isLoading: PropTypes.bool
};

export default Details;
