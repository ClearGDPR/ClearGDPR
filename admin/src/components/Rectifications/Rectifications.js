import React from 'react';
import PropTypes from 'prop-types';

import { Tab, Tabs, TabList, TabPanel } from 'react-tabs';
import 'react-tabs/style/react-tabs.css';
import 'theme/Tabs.css';

import Table from 'components/core/Common/Table/Table';
import Loader from 'components/core/cards/dashboard/Loader';
import Card from 'components/core/cards/dashboard/Card';
import Paginate from 'components/core/Paginate';
import ActionBar from 'components/core/Common/Bars/ActionBar';

const actions = ({ onDetailsClick }) => [
  {
    label: 'Details',
    onClick: (e, item) => {
      e.preventDefault();
      onDetailsClick(item.id);
    }
  }
];

const columns = ({ showStatus } = { showStatus: false }) => ({
  id: {
    title: 'ID'
  },
  created_at: {
    title: 'Created at'
  },
  request_reason: {
    title: 'Reason'
  },
  status: {
    title: 'Status',
    filter: () => showStatus
  }
});

const Rectifications = ({
  isLoading,
  tabs,
  selectedTab,
  onTabSelect,
  pageCount,
  currentPage,
  onPageSelected,
  onDetailsClick,
  data
}) => {
  function handlePageClick(page) {
    onPageSelected(page);
  }

  function renderPagination() {
    return (
      <Paginate
        pageCount={pageCount}
        initialPage={currentPage - 1}
        onPageChange={({ selected }) => handlePageClick(selected + 1)}
        disableInitialCallback={true}
      />
    );
  }

  function renderPendingRequests() {
    return (
      <React.Fragment>
        <div className="content">
          {data.length > 0 ? (
            <Table rows={data} columns={columns()} actions={actions({ onDetailsClick })} />
          ) : (
            <i>No pending requests registered</i>
          )}
        </div>
        <div className="content">{renderPagination()}</div>
      </React.Fragment>
    );
  }

  function renderRequestsArchive() {
    return (
      <React.Fragment>
        <div className="content">
          {data.length > 0 ? (
            <Table
              rows={data}
              columns={columns({ showStatus: true })}
              actions={actions({ onDetailsClick })}
            />
          ) : (
            <i>No requests archieved</i>
          )}
        </div>
        <div className="content">{renderPagination()}</div>
      </React.Fragment>
    );
  }

  function renderLoader() {
    return (
      <div className="content">
        <Loader />
      </div>
    );
  }

  return (
    <section className="cards">
      <ActionBar
        title="Rectification requests"
        desc="Manage requests from users who want to execute their right to rectify the
            data stored about them."
      />
      <Tabs selectedIndex={selectedTab} onSelect={onTabSelect}>
        <TabList>{tabs.map((value, index) => <Tab key={index}>{value}</Tab>)}</TabList>
        <TabPanel>
          <div className="row">
            <Card cols={8}>{!isLoading ? renderPendingRequests() : renderLoader()}</Card>
          </div>
        </TabPanel>
        <TabPanel>
          <div className="row">
            <Card cols={8}>{!isLoading ? renderRequestsArchive() : renderLoader()}</Card>
          </div>
        </TabPanel>
      </Tabs>
    </section>
  );
};

Rectifications.propTypes = {
  isLoading: PropTypes.bool,
  tabs: PropTypes.array.isRequired,
  selectedTab: PropTypes.number.isRequired,
  onTabSelect: PropTypes.func.isRequired,
  pageCount: PropTypes.number.isRequired,
  currentPage: PropTypes.number.isRequired,
  onPageSelected: PropTypes.func.isRequired,
  onDetailsClick: PropTypes.func.isRequired,
  data: PropTypes.array.isRequired
};

export default Rectifications;
