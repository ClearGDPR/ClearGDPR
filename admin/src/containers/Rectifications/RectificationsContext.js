import React, { Component, createContext } from 'react';
import PropTypes from 'prop-types';
import config from 'config';
import internalFetch from 'helpers/internal-fetch';
import { toast } from 'react-toastify';
import { format } from 'date-fns';

const RectificationsContext = createContext({
  pendingRectifications: {},
  processedRectifications: {},
  fetchPendingRectifications: () => {},
  fetchProcessedRectifications: () => {},
  fetchAllRectifications: () => {},
  isLoading: false
});

export class RectificationsProvider extends Component {
  static propTypes = {
    children: PropTypes.oneOfType([
      PropTypes.element,
      PropTypes.object,
      PropTypes.arrayOf(PropTypes.element)
    ])
  };

  state = {
    pendingRectifications: {},
    processedRectifications: {},
    fetchPendingRectifications: this.fetchPendingRectifications.bind(this),
    fetchProcessedRectifications: this.fetchProcessedRectifications.bind(this),
    fetchAllRectifications: this.fetchAllRectifications.bind(this),
    isLoading: false
  };

  async _getPendingRectifications(page = 1) {
    return this._mapRectificationResults(
      await internalFetch(
        `${config.API_URL}/api/management/subjects/rectification-requests/list?page=${page}`
      )
    );
  }

  async _getProcessedRectifications(page = 1) {
    return this._mapRectificationResults(
      await internalFetch(
        `${config.API_URL}/api/management/subjects/rectification-requests/archive?page=${page}`
      )
    );
  }

  _mapRectificationResults(result) {
    let mappedData = result.data.map(v => {
      let mapped = {
        id: v.id,
        created_at: format(new Date(v.created_at), 'DD/MM/YYYY h:mma'),
        request_reason: v.request_reason
      };

      v.status && (mapped.status = v.status);
      return mapped;
    });
    return {
      ...result,
      data: mappedData
    };
  }

  setLoading(loading) {
    this.setState({
      isLoading: loading
    });
  }

  async cancelLoadingAndReject(e) {
    this.setLoading(false);
    toast.error(`An error occurred: ${e.message}`);
  }

  async fetchPendingRectifications(page = 1) {
    await this._fetchRectifications(false, page);
  }

  async fetchProcessedRectifications(page = 1) {
    await this._fetchRectifications(true, page);
  }

  async fetchAllRectifications(page = 1) {
    await this._fetchRectifications(null, page);
  }

  async _fetchRectifications(archive = null, page = 1) {
    this.setLoading(true);

    try {
      const pendingRectifications =
        archive === null || !archive
          ? await this._getPendingRectifications(page)
          : this.state.pendingRectifications;
      const processedRectifications =
        archive === null || archive
          ? await this._getProcessedRectifications(page)
          : this.state.processedRectifications;
      this.setState({
        pendingRectifications,
        processedRectifications,
        isLoading: false
      });
    } catch (e) {
      await this.cancelLoadingAndReject(e);
    }
  }

  render() {
    return (
      <RectificationsContext.Provider value={this.state}>
        {this.props.children}
      </RectificationsContext.Provider>
    );
  }
}

export const RectificationsConsumer = RectificationsContext.Consumer;
