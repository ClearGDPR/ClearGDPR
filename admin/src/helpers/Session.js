import cookie from 'cookie';

export class Session {
  set(authResult) {
    const maxAge = 1 * 24 * 60 * 60; // 1 day long;
    document.cookie = cookie.serialize('auth', authResult, { maxAge });
    document.cookie = cookie.serialize('expires_at', authResult, { maxAge });
  }

  destroy() {
    document.cookie = cookie.serialize('auth', '', { maxAge: -1 });
    document.cookie = cookie.serialize('expires_at', '', { maxAge: -1 });
  }

  isLoggedIn() {
    let expiresAt = cookie.parse(document.cookie).expires_at || '';
    return new Date().getTime() < expiresAt;
  }
}

export default new Session();
