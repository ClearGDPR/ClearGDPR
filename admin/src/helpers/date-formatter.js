import { format } from 'date-fns';

export default date => {
  return format(new Date(date), 'DD/MM/YYYY h:mma');
};
