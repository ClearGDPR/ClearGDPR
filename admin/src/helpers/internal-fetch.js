import merge from 'lodash/object/merge';
import session from '../../helpers/Session';

const internalFetch = (url, config = {}) => {
  const internalConfig = merge(
    {
      method: 'GET',
      headers: {
        Authorization: `Bearer ${session.getToken()}`,
        'Content-Type': 'application/json'
      }
    },
    config
  );
  return fetch(url, internalConfig).then(async res => {
    if (res.status === 400 && (await res.json()).error === 'JWT token expired') {
      session.destroy();
    } else {
      return res;
    }
  });
};

export default internalFetch;
