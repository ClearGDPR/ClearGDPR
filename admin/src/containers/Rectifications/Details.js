import React from 'react';
import PropTypes from 'prop-types';
import format from 'helpers/date-formatter';

import config from 'config';
import internalFetch from 'helpers/internal-fetch';

import { RectificationsConsumer } from './RectificationsContext';
import { PanelConsumer } from 'containers/MainLayout/PanelContext';
import Details from 'components/Rectifications/Details';

export class DetailsContainer extends React.Component {
  static propTypes = {
    approveRectification: PropTypes.func,
    rectificationId: PropTypes.number,
    closePanel: PropTypes.func,
    isLoading: PropTypes.bool
  };

  state = {
    rectification: {}
  };

  componentDidMount() {
    this.setState({
      isLoading: true
    });

    internalFetch(
      `${config.API_URL}/api/management/subjects/rectification-requests/${
        this.props.rectificationId
      }`
    )
      .then(r =>
        this.setState({
          rectification: {
            created_at: format(r.createdAt),
            currentData: r.currentData,
            updates: r.updates,
            status: r.status
          },
          isLoading: false
        })
      )
      .catch(() => this.setState({ isLoading: false }));
  }

  get isLoading() {
    return this.state.isLoading || this.props.isLoading;
  }

  onApprove = () => {
    this.setState({
      isLoading: true
    });

    this.props
      .approveRectification(this.props.rectificationId)
      .then(() => this.props.closePanel())
      .catch(() => {
        this.setState({
          isLoading: false
        });
      });
  };

  render() {
    return (
      <Details
        rectification={this.state.rectification}
        onApprove={this.onApprove}
        isLoading={this.isLoading}
      />
    );
  }
}

export default props => (
  <PanelConsumer>
    {({ closePanel }) => (
      <RectificationsConsumer>
        {({ approveRectification, isLoading }) => (
          <DetailsContainer
            {...props}
            isLoading={isLoading}
            approveRectification={approveRectification}
            closePanel={closePanel}
          />
        )}
      </RectificationsConsumer>
    )}
  </PanelConsumer>
);
