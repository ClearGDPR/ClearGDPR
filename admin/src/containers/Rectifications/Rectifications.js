import React from 'react';
import PropTypes from 'prop-types';

import { RectificationsConsumer } from 'containers/Rectifications/RectificationsContext';
import { PanelConsumer } from 'containers/MainLayout/PanelContext';

import Rectifications from 'components/Rectifications/Rectifications';
import Details from 'components/Rectifications/Details';

export class RectificationsContainer extends React.Component {
  static propTypes = {
    pendingRectifications: PropTypes.object.isRequired,
    processedRectifications: PropTypes.object.isRequired,
    fetchPendingRectifications: PropTypes.func.isRequired,
    fetchProcessedRectifications: PropTypes.func.isRequired,
    fetchAllRectifications: PropTypes.func.isRequired,
    isLoading: PropTypes.bool.isRequired,
    openPanel: PropTypes.func.isRequired
  };

  state = {
    selectedTab: 0,
    tabs: ['Pending requests', 'Requests archive']
  };

  get pendingTabActive() {
    return this.state.selectedTab === 0;
  }

  onTabSelect(tabIndex) {
    this.setState({
      selectedTab: tabIndex
    });
  }

  getRectificationsData() {
    if (this.pendingTabActive) {
      return this.props.pendingRectifications;
    } else {
      return this.props.processedRectifications;
    }
  }

  getData() {
    return this.getRectificationsData().data || [];
  }

  componentDidMount() {
    this.props.fetchAllRectifications();
  }

  getPageCount() {
    const paging = this.getRectificationsData().paging;
    return paging ? paging.total : 1;
  }

  getCurrentPage() {
    const paging = this.getRectificationsData().paging;
    return paging ? paging.current : 1;
  }

  onPageSelected(page) {
    if (this.pendingTabActive) {
      this.props.fetchPendingRectifications(page);
    } else {
      this.props.fetchProcessedRectifications(page);
    }
  }

  onDetailsClick(id) {
    this.props.openPanel(Details, 'Rectification details', { rectificationId: id });
  }

  render() {
    return (
      <Rectifications
        tabs={this.state.tabs}
        selectedTab={this.state.selectedTab}
        onTabSelect={this.onTabSelect.bind(this)}
        pageCount={this.getPageCount()}
        currentPage={this.getCurrentPage()}
        data={this.getData()}
        isLoading={this.props.isLoading}
        onPageSelected={this.onPageSelected.bind(this)}
        onDetailsClick={this.onDetailsClick.bind(this)}
      />
    );
  }
}

export default props => (
  <PanelConsumer>
    {({ openPanel }) => (
      <RectificationsConsumer>
        {({
          pendingRectifications,
          processedRectifications,
          fetchPendingRectifications,
          fetchProcessedRectifications,
          fetchAllRectifications,
          isLoading
        }) => (
          <RectificationsContainer
            {...props}
            pendingRectifications={pendingRectifications}
            processedRectifications={processedRectifications}
            fetchPendingRectifications={fetchPendingRectifications}
            fetchProcessedRectifications={fetchProcessedRectifications}
            fetchAllRectifications={fetchAllRectifications}
            isLoading={isLoading}
          />
        )}
      </RectificationsConsumer>
    )}
  </PanelConsumer>
);
