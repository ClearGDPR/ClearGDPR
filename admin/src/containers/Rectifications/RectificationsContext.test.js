import React from 'react';

import { shallow } from 'enzyme';
import { toast } from 'react-toastify';
import { format } from 'date-fns';

import internalFetch from 'helpers/internal-fetch';
import * as TestUtils from 'tests/helpers/TestUtils';

import { RectificationsProvider } from '../Rectifications/RectificationsContext';
import * as lodash from 'lodash';

jest.mock('date-fns');
jest.mock('react-toastify');
jest.mock('helpers/internal-fetch');

beforeAll(() =>
  format.mockImplementation(date => {
    return date.toUTCString();
  }));

beforeEach(() => {
  jest.clearAllMocks();
});

describe('RectificationsProvider', () => {
  const pendingRectifications = {
    data: [
      {
        id: 1,
        request_reason: 'The data was incorrect.',
        created_at: '2018-07-02T21:31:24.999Z'
      },
      {
        id: 2,
        request_reason: 'The data was incorrect two.',
        created_at: '2018-07-02T21:31:37.440Z'
      },
      {
        id: 3,
        request_reason: 'The data was incorrect two three.',
        created_at: '2018-07-02T21:31:43.530Z'
      }
    ],
    paging: {
      current: 1,
      total: 2
    }
  };

  const processedRectifications = {
    data: [
      {
        id: 4,
        request_reason: 'The data was incorrect.',
        created_at: '2018-07-02T21:31:24.999Z',
        status: 'APPROVED'
      },
      {
        id: 5,
        request_reason: 'The data was incorrect two.',
        created_at: '2018-07-02T21:31:37.440Z',
        status: 'DISAPPROVED'
      },
      {
        id: 6,
        request_reason: 'The data was incorrect two three.',
        created_at: '2018-07-02T21:31:43.530Z',
        status: 'APPROVED'
      }
    ],
    paging: {
      current: 1,
      total: 2
    }
  };

  const setup = propOverrides => {
    const props = Object.assign({}, propOverrides);
    const component = shallow(<RectificationsProvider {...props} />);

    return { props, component };
  };

  it('should render correctly', async () => {
    internalFetch.mockReturnValue(Promise.resolve(pendingRectifications));

    const { component } = setup();

    await TestUtils.flushPromises();
    expect(component).toMatchSnapshot();
  });

  describe('fetching rectifications', () => {
    it('should have correct state after fetching all', async () => {
      let i = 0;
      internalFetch.mockImplementation(() =>
        Promise.resolve(0 === i++ ? pendingRectifications : processedRectifications)
      );

      const { component } = setup();
      await component.instance().fetchAllRectifications();

      await TestUtils.flushPromises();

      let state = component.state();
      expect(state).toEqual(
        expect.objectContaining({
          isLoading: false
        })
      );
      expect(state.pendingRectifications).toMatchSnapshot();
      expect(state.processedRectifications).toMatchSnapshot();
    });

    it('should have correct state after fetching pending', async () => {
      internalFetch.mockReturnValue(Promise.resolve(pendingRectifications));

      const { component } = setup();
      await component.instance().fetchPendingRectifications();

      await TestUtils.flushPromises();

      let state = component.state();
      expect(state).toEqual(
        expect.objectContaining({
          processedRectifications: { paging: { current: 1 } },
          isLoading: false
        })
      );
      expect(state.pendingRectifications).toMatchSnapshot();
    });

    it('should have correct state after fetching processed', async () => {
      internalFetch.mockReturnValue(Promise.resolve(processedRectifications));

      const { component } = setup();
      await component.instance().fetchProcessedRectifications();

      await TestUtils.flushPromises();

      let state = component.state();
      expect(state).toEqual(
        expect.objectContaining({
          pendingRectifications: { paging: { current: 1 } },
          isLoading: false
        })
      );
      expect(state.processedRectifications).toMatchSnapshot();
    });

    it('should have correct state when calling API returns error', async () => {
      internalFetch.mockReturnValue(Promise.reject(new Error('Error message')));

      const { component } = setup();

      try {
        await component.instance().fetchAllRectifications();
      } catch (e) {
        //not-empty
      }

      expect(component.state()).toEqual(
        expect.objectContaining({
          isLoading: false
        })
      );
    });
  });

  describe('approve rectification', () => {
    it('should refresh rectifications', async () => {
      const newPending = {
        ...pendingRectifications,
        data: pendingRectifications.data.filter(r => r.id !== 1)
      };
      const newProcessed = {
        ...processedRectifications,
        data: [
          ...processedRectifications.data,
          {
            id: 1,
            request_reason: 'The data was incorrect.',
            created_at: '2018-07-02T21:31:24.999Z'
          }
        ]
      };

      internalFetch.mockImplementation(url => {
        if (lodash.includes(url, 'rectification-requests/archive')) {
          return Promise.resolve(newProcessed);
        } else if (lodash.includes(url, 'rectification-requests')) {
          return Promise.resolve(newPending);
        } else {
          return Promise.resolve({ success: true });
        }
      });

      const { component } = setup();

      await component.instance().approveRectification(973);

      const state = component.state();
      expect(state).toEqual(
        expect.objectContaining({
          isLoading: false,
          pendingRectifications: component.instance()._mapRectificationResults(newPending),
          processedRectifications: component.instance()._mapRectificationResults(newProcessed)
        })
      );
      expect(toast.success).toHaveBeenCalledWith('Rectification request approved');
    });

    it('should stop loading when approval fails', async () => {
      internalFetch.mockReturnValue(Promise.reject(new Error('Error message')));

      const { component } = setup();

      try {
        await component.instance().approveRectification(2);
      } catch (e) {
        //not-empty
      }

      expect(component.state()).toEqual(
        expect.objectContaining({
          isLoading: false
        })
      );
    });
  });
});
