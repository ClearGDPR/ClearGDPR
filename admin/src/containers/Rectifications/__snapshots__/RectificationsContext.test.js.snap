// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RectificationsProvider should have correct state after fetching all rectifications 1`] = `
Object {
  "data": Array [
    Object {
      "created_at": "Mon, 02 Jul 2018 21:31:24 GMT",
      "id": 1,
      "request_reason": "The data was incorrect.",
    },
    Object {
      "created_at": "Mon, 02 Jul 2018 21:31:37 GMT",
      "id": 2,
      "request_reason": "The data was incorrect two.",
    },
    Object {
      "created_at": "Mon, 02 Jul 2018 21:31:43 GMT",
      "id": 3,
      "request_reason": "The data was incorrect two three.",
    },
  ],
  "paging": Object {
    "current": 1,
    "total": 2,
  },
}
`;

exports[`RectificationsProvider should have correct state after fetching all rectifications 2`] = `
Object {
  "data": Array [
    Object {
      "created_at": "Mon, 02 Jul 2018 21:31:24 GMT",
      "id": 4,
      "request_reason": "The data was incorrect.",
      "status": "APPROVED",
    },
    Object {
      "created_at": "Mon, 02 Jul 2018 21:31:37 GMT",
      "id": 5,
      "request_reason": "The data was incorrect two.",
      "status": "DISAPPROVED",
    },
    Object {
      "created_at": "Mon, 02 Jul 2018 21:31:43 GMT",
      "id": 6,
      "request_reason": "The data was incorrect two three.",
      "status": "APPROVED",
    },
  ],
  "paging": Object {
    "current": 1,
    "total": 2,
  },
}
`;

exports[`RectificationsProvider should have correct state after fetching pending rectifications 1`] = `
Object {
  "data": Array [
    Object {
      "created_at": "Mon, 02 Jul 2018 21:31:24 GMT",
      "id": 1,
      "request_reason": "The data was incorrect.",
    },
    Object {
      "created_at": "Mon, 02 Jul 2018 21:31:37 GMT",
      "id": 2,
      "request_reason": "The data was incorrect two.",
    },
    Object {
      "created_at": "Mon, 02 Jul 2018 21:31:43 GMT",
      "id": 3,
      "request_reason": "The data was incorrect two three.",
    },
  ],
  "paging": Object {
    "current": 1,
    "total": 2,
  },
}
`;

exports[`RectificationsProvider should have correct state after fetching processed rectifications 1`] = `
Object {
  "data": Array [
    Object {
      "created_at": "Mon, 02 Jul 2018 21:31:24 GMT",
      "id": 4,
      "request_reason": "The data was incorrect.",
      "status": "APPROVED",
    },
    Object {
      "created_at": "Mon, 02 Jul 2018 21:31:37 GMT",
      "id": 5,
      "request_reason": "The data was incorrect two.",
      "status": "DISAPPROVED",
    },
    Object {
      "created_at": "Mon, 02 Jul 2018 21:31:43 GMT",
      "id": 6,
      "request_reason": "The data was incorrect two three.",
      "status": "APPROVED",
    },
  ],
  "paging": Object {
    "current": 1,
    "total": 2,
  },
}
`;

exports[`RectificationsProvider should render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <RectificationsProvider />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": undefined,
      "value": Object {
        "fetchAllRectifications": [Function],
        "fetchPendingRectifications": [Function],
        "fetchProcessedRectifications": [Function],
        "isLoading": false,
        "pendingRectifications": Object {},
        "processedRectifications": Object {},
      },
    },
    "ref": null,
    "rendered": null,
    "type": Object {
      "$$typeof": Symbol(react.provider),
      "_context": Object {
        "$$typeof": Symbol(react.context),
        "Consumer": [Circular],
        "Provider": [Circular],
        "_calculateChangedBits": null,
        "_changedBits": 0,
        "_changedBits2": 0,
        "_currentRenderer": null,
        "_currentRenderer2": null,
        "_currentValue": Object {
          "fetchAllRectifications": [Function],
          "fetchPendingRectifications": [Function],
          "fetchProcessedRectifications": [Function],
          "isLoading": false,
          "pendingRectifications": Object {},
          "processedRectifications": Object {},
        },
        "_currentValue2": Object {
          "fetchAllRectifications": [Function],
          "fetchPendingRectifications": [Function],
          "fetchProcessedRectifications": [Function],
          "isLoading": false,
          "pendingRectifications": Object {},
          "processedRectifications": Object {},
        },
        "_defaultValue": Object {
          "fetchAllRectifications": [Function],
          "fetchPendingRectifications": [Function],
          "fetchProcessedRectifications": [Function],
          "isLoading": false,
          "pendingRectifications": Object {},
          "processedRectifications": Object {},
        },
      },
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": undefined,
        "value": Object {
          "fetchAllRectifications": [Function],
          "fetchPendingRectifications": [Function],
          "fetchProcessedRectifications": [Function],
          "isLoading": false,
          "pendingRectifications": Object {},
          "processedRectifications": Object {},
        },
      },
      "ref": null,
      "rendered": null,
      "type": Object {
        "$$typeof": Symbol(react.provider),
        "_context": Object {
          "$$typeof": Symbol(react.context),
          "Consumer": [Circular],
          "Provider": [Circular],
          "_calculateChangedBits": null,
          "_changedBits": 0,
          "_changedBits2": 0,
          "_currentRenderer": null,
          "_currentRenderer2": null,
          "_currentValue": Object {
            "fetchAllRectifications": [Function],
            "fetchPendingRectifications": [Function],
            "fetchProcessedRectifications": [Function],
            "isLoading": false,
            "pendingRectifications": Object {},
            "processedRectifications": Object {},
          },
          "_currentValue2": Object {
            "fetchAllRectifications": [Function],
            "fetchPendingRectifications": [Function],
            "fetchProcessedRectifications": [Function],
            "isLoading": false,
            "pendingRectifications": Object {},
            "processedRectifications": Object {},
          },
          "_defaultValue": Object {
            "fetchAllRectifications": [Function],
            "fetchPendingRectifications": [Function],
            "fetchProcessedRectifications": [Function],
            "isLoading": false,
            "pendingRectifications": Object {},
            "processedRectifications": Object {},
          },
        },
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
